MediaBrowser.MediaEncoding/Encoder/MediaEncoder.cs

```diff
@@ -32,323 +28,288 @@ namespace MediaBrowser.MediaEncoding.Encoder
-            }
-
-            path = newPaths.Item1;
+                else
+                {
+                    // Supplied path could be either file path or folder path.
+                    // Resolve down to file path and validate
+                    path = GetEncoderPath(path);
 
-            if (!ValidateVersion(path, true))
-            {
-                throw new ResourceNotFoundException("ffmpeg version 3.0 or greater is required.");
-            }
+                    if (path == null)
+                    {
+                        throw new ResourceNotFoundException("FFmpeg not found");
+                    }
+                    else if (!ValidatePathFFmpeg("New From UI", path))
+                    {
+                        throw new ResourceNotFoundException("Failed validation checks.  Version 4.0 or greater is required");
+                    }
+                    else
+                    {
+                        EncoderLocationType = "Custom";
 
-            var config = GetEncodingOptions();
-            config.EncoderAppPath = path;
-            ConfigurationManager.SaveConfiguration("encoding", config);
+                        // Write the validated mpeg path to the xml as <EncoderAppPathCustom>
+                        // This ensures its not lost on new startup
+                        var config = ConfigurationManager.GetConfiguration<EncodingOptions>("encoding");
+                        config.EncoderAppPathCustom = FFmpegPath;
+                        ConfigurationManager.SaveConfiguration("encoding", config);
 
-            Init();
-        }
+                        FFprobePath = null; // Clear probe path so it gets relooked in ReInit()
 
-        private bool ValidateVersion(string path, bool logOutput)
-        {
-            return new EncoderValidator(_logger, _processFactory).ValidateVersion(path, logOutput);
+                        ReInit();
+                    }
+                }
+            }
         }
 
-        private void ConfigureEncoderPaths()
+        private bool ValidatePath(string type, string path)
         {
-            if (_hasExternalEncoder)
+            if (!string.IsNullOrEmpty(path))
             {
-                return;
-            }
-
-            var appPath = GetEncodingOptions().EncoderAppPath;
+                if (File.Exists(path))
+                {
+                    var valid = new EncoderValidator(_logger, _processFactory).ValidateVersion(path, true);
 
-            if (string.IsNullOrWhiteSpace(appPath))
-            {
-                appPath = Path.Combine(ConfigurationManager.ApplicationPaths.ProgramDataPath, "ffmpeg");
+                    if (valid == true)
+                    {
+                        return true;
+                    }
+                    else
+                    {
+                        _logger.LogError("{0}: Failed validation checks.  Version 4.0 or greater is required: {1}", type, path);
```

Comment:
This message is wrong in terms of it contains specific `ffmpeg` version (which is subject to change in the future).
It should be written by `ValidateVersion` (i.e. in the place where "good" ffmpeg version is known).

Reply:
Ok good point.